falcon/app_helpers.py

```diff
@@ -311,13 +310,27 @@ def default_serialize_error(req: Request, resp: Response, exception: HTTPError)
             preferred = MEDIA_JSON
         elif '+xml' in accept:
             preferred = MEDIA_XML
+        else:
+            # NOTE(caselit): if nothing else matchers try using the media handlers
+            # registered in the response
+            preferred = req.client_prefers(list(resp.options.media_handlers))
 
     if preferred is not None:
+        handler, serialize_sync, _ = resp.options.media_handlers._resolve(
+            preferred, MEDIA_JSON, raise_not_found=False
+        )
         if preferred == MEDIA_JSON:
-            handler, _, _ = resp.options.media_handlers._resolve(
-                MEDIA_JSON, MEDIA_JSON, raise_not_found=False
-            )
+            # NOTE(caselit): special case json to ensure that it's always possible to
+            # serialize an error in json even if no handler is set in the
+            # media_handlers.
             resp.data = exception.to_json(handler)
+        elif handler:
+            if serialize_sync:
```

Comment:
@vytas7 not sure we need to bother with this optimization here.

Reply:
I don't think we need this path, there is already the same optimization in place for rendering out the body from media. Just set the media and let the app do this.

Reply:
I agree. Will simplify